### PR TITLE
Determine ptable dynamically

### PR DESCRIPTION
**Pull request by fritzmg**
Currently you have to do something like

```php
// src/Resources/contao/dca/tl_content.php
if (\Contao\Input::get('do') == 'foo')
{
	$GLOBALS['TL_DCA']['tl_content']['config']['ptable'] = 'tl_foo';
}
```

in order to be able to use content elements for your own back end module's table. With this PR the parent table will be determined dynamically according to the back end module configurations.

### CODE REVIEW

**Comment by fritzmg** [comment]
Failing test looks unrelated to this PR (use of undefined constant `TL_MODE` in `core-bundle\src\Controller\BackendCsvImportController.php`).

File: core-bundle/src/Resources/contao/library/Contao/Controller.php [hunk 1]
```diff
@@ -1384,7 +1384,7 @@ protected function redirectToFrontendPage($intPage, $strArticle=null, $blnReturn
 	protected function getParentEntries($strTable, $intId)
 	{
 		// No parent table
-		if (!isset($GLOBALS['TL_DCA'][$strTable]['config']['ptable']))
+		if (!isset($GLOBALS['TL_DCA'][$strTable]['config']['ptable']) || empty($GLOBALS['TL_DCA'][$strTable]['config']['ptable']))
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Controller.php]:
```suggestion
		if (empty($GLOBALS['TL_DCA'][$strTable]['config']['ptable']))
```

I'm pretty sure that `empty()` implies the `isset()` check.

**Review by leofeyer — CHANGES_REQUESTED** [review]
Maybe we can reduce the complexity of the `setDynamicPTable()` method with some early returns?

**Review by bytehead — DISMISSED** [review]

File: core-bundle/src/Resources/contao/library/Contao/DcaLoader.php [hunk 2]
```diff
@@ -198,6 +201,39 @@ private function addDefaultLabels($blnNoCache)
...
+			return;
+		}
+
+		if (null !== ($do = Input::get('do')))
+		{
+			foreach (array_merge(...array_values($GLOBALS['BE_MOD'])) as $key => $module)
+			{
+				if ($do === $key)
+				{
+					foreach ($module['tables'] ?? array() as $table)
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/library/Contao/DcaLoader.php]:
```suggestion
				if ($do === $key && isset($module['tables']) && is_array($module['tables']))
				{
					foreach ($module['tables'] as $table)
```

**Review by aschempp — COMMENTED** [review]
Interesting approach! I wonder if we ever need to set `ptable` at all, because it could be determined by this statement (regardless of being `dynamicPtable`)?

**Comment by fritzmg** [comment]
> Interesting approach! I wonder if we ever need to set `ptable` at all, because it could be determined by this statement (regardless of being `dynamicPtable`)?

In theory. But performance-wise it would be better to define it.

**Comment by Toflar** [comment]
Is there any reason why the current code is not adjusted? I mean the one in news, calendar etc.

**Comment by fritzmg** [comment]
> Is there any reason why the current code is not adjusted? I mean the one in news, calendar etc.

I only did not remove it, because there are still other adjustments that the news-bundle makes within `tl_content.php`. And then it is still more efficient to directly set the ptable anyway. However, as we found out at the meeting, these two can be removed:

```php
array_unshift($GLOBALS['TL_DCA']['tl_content']['config']['onload_callback'], array('tl_content_news', 'checkPermission'));
$GLOBALS['TL_DCA']['tl_content']['list']['operations']['toggle']['button_callback'] = array('tl_content_news', 'toggleIcon');
```

(needs to be a bugfix for `4.9` actually)

So the only remaining line is

```php
$GLOBALS['TL_DCA']['tl_content']['config']['onload_callback'][] = array('tl_content_news', 'generateFeed');
```

But I think this can be changed too, so that the `onload_callback` is always loaded, and within it, you simply check whether the `ptable` is `tl_news`.

Overall I would do that in a follow-up PR, if this gets merged.

File: core-bundle/src/Resources/contao/library/Contao/DcaLoader.php [hunk 3]
```diff
@@ -198,6 +201,39 @@ private function addDefaultLabels($blnNoCache)
...
+	 * Sets the parent table for the current table, if enabled and not set.
+	 */
+	private function setDynamicPTable(): void
+	{
+		if (!($GLOBALS['TL_DCA'][$this->strTable]['config']['dynamicPtable'] ?? false) || !empty($GLOBALS['TL_DCA'][$this->strTable]['config']['ptable']) || !isset($GLOBALS['BE_MOD']))
+		{
+			return;
+		}
+
+		if (null !== ($do = Input::get('do')))
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/DcaLoader.php]: This should be an early return.

```suggestion
		if (!$do = Input::get('do'))
		{
			return;
		}
```

File: core-bundle/src/Resources/contao/library/Contao/DcaLoader.php [hunk 4]
```diff
@@ -198,6 +201,39 @@ private function addDefaultLabels($blnNoCache)
...
+		{
+			foreach (array_merge(...array_values($GLOBALS['BE_MOD'])) as $key => $module)
+			{
+				if ($do === $key && isset($module['tables']) && \is_array($module['tables']))
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/DcaLoader.php]: This should be an early return, too.

```suggestion
				if ($do !== $key || !isset($module['tables']) || !\is_array($module['tables']))
				{
					continue;
				}
```

**Review by leofeyer — CHANGES_REQUESTED** [review]

**Review by leofeyer — COMMENTED** [review]
Looks good. 👍 Can/should we remove `'dynamicPtable' => true,` from the tl_content DCA file?

**Comment by fritzmg** [comment]
> Can/should we remove `'dynamicPtable' => true,` from the tl_content DCA file?

No, that's still necessary. What's not necessary anymore is doing this in your own bundle:

```php
// src/Resources/contao/dca/tl_content.php
if (\Contao\Input::get('do') == 'foo') {
	$GLOBALS['TL_DCA']['tl_content']['config']['ptable'] = 'tl_foo';
}
```

**Review by bytehead — APPROVED** [review]

**Review by ausi — APPROVED** [review]

**Comment by leofeyer** [comment]
Ok, then we should remove `$GLOBALS['TL_DCA']['tl_content']['config']['ptable'] = '…';` in the following files, shouldn't we?

* `calendar-bundle/src/Resources/contao/dca/tl_content.php`
* `core-bundle/src/Resources/contao/dca/tl_content.php`
* `news-bundle/src/Resources/contao/dca/tl_content.php`

**Comment by ausi** [comment]
Wouldn’t that be a BC break then if someone reads from `$GLOBALS['TL_DCA']['tl_content']['config']['ptable']` ?

**Comment by leofeyer** [comment]
I don't think so, as the value is re-added dynamically, isn't it?

**Comment by fritzmg** [comment]
> I don't think so, as the value is re-added dynamically, isn't it?

For callbacks, that's true. However, if anyone is relying on `$GLOBALS['TL_DCA']['tl_content']['config']['ptable']` being present in `contao/dca/tl_content.php` of their application or extension, then it would be missing there, since it is of course only added after all DCA files have been loaded.

**Comment by fritzmg** [comment]
> And then it is still more efficient to directly set the ptable anyway. However, as we found out at the meeting, these two can be removed:
> 
> ```php
> array_unshift($GLOBALS['TL_DCA']['tl_content']['config']['onload_callback'], array('tl_content_news', 'checkPermission'));
> ```

I was wrong about this one. While `tl_content_news::checkPermission` _almost_ identical to `tl_content::checkPermission`, it does have a few adjustments of course.

**Review by aschempp — APPROVED** [review]
I think the feature and implementation is „perfect“. But should we consider setting it in a loadDataContainer hook instead of DC_Table? On the other hand, the value can‘t be present e.g. in the front end anyway 🤔

**Comment by leofeyer** [comment]
> the value can‘t be present e.g. in the front end anyway 🤔

Why not?

**Comment by fritzmg** [comment]
Hmm... what would be the advantage of having it in the hook instead?

**Review by leofeyer — APPROVED** [review]
This PR is ready to be merged from my point of view. Let's wait for @aschempp to elaborate.

**Comment by aschempp** [comment]
> > the value can‘t be present e.g. in the front end anyway 🤔
> 
> Why not?

because the `ptable` is based on the current back end module?

**Comment by leofeyer** [comment]
As briefly discussed in Slack:

> The `onload_callback` will only set the ptable if `new DC_Table()` is run. The hook would set the ptable anywhere when `Controller::loadDataContainer()` is executed.

That is a good point actually. Are we 100% sure that we only need the ptable in the back end? Because I am not…

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Comment by fritzmg** [comment]
For anyone interested: a backport of this functionality is also available now for Contao 4.9: https://packagist.org/packages/inspiredminds/contao-dynamic-ptable-backport

**Pull request merged**